study1/week3/evaluate-the-bracket-pairs-of-a-string.java

```diff
@@ -0,0 +1,26 @@
+// https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string
+// Time Complexity: O(n)
+// Space Complexity: O(n), because of hashmap
+class Solution {
+    public String evaluate(String s, List<List<String>> knowledge) {
+        String result = "";
+        HashMap<String, String> wordMap = new HashMap<>();  // key: word to be changed, value: changing word
+
+        for (List<String> word : knowledge) {
+            wordMap.put("(" + word.get(0) + ")", word.get(1));
+        }
+
+        for (int i = 0; i < s.length(); ++i) {
+            char ch = s.charAt(i);
+            if (ch == '(') {    // bracket found
+                int end = s.indexOf(")", i);
+                result += wordMap.getOrDefault(s.substring(i, end+1), "?");
```

Comment:
This is a nice use of getOoDefault.

One other thing is that if  you would have done the following instead,
```java
result += wordMap.getOrDefault(s.substring(i + 1, end), "?");
```
you didn't need to concatenate the brackets in line 10. this will save few computations.